src/jgikbase/idmapping/storage/mongo/id_mapping_mongo_storage.py

```diff
@@ -1,22 +1,73 @@
 """
-
-@author: gaprice@lbl.gov
+A MongoDB based storage system for ID mapping.
 """
 from jgikbase.idmapping.storage.id_mapping_storage import IDMappingStorage as _IDMappingStorage
-from jgikbase.idmapping.core.object_id import NamespaceID
+from jgikbase.idmapping.core.tokens import HashedToken
+from jgikbase.idmapping.core.user import User, AuthsourceID
+from pymongo.database import Database
+from jgikbase.idmapping.util.util import not_none
+
+# TODO NOW implement remaining methods in superclass
+# TODO NOW implement database schema checking
+# TODO NOW test indexes are correct
+# TODO NOW finish tests
+
+# Testing the (many) catch blocks for the general mongo exception is pretty hard, since it
+# appears as though the mongo clients have a heartbeat, so just stopping mongo might trigger
+# the heartbeat exception rather than the exception you're going for.
+
+# Mocking the mongo client is probably not the answer:
+# http://stackoverflow.com/questions/7413985/unit-testing-with-mongodb
+# https://github.com/mockito/mockito/wiki/How-to-write-good-tests
+
+_COL_USERS = 'users'
 
-# TODO NOW
+_FLD_AUTHSOURCE = 'auth'
+_FLD_USER = 'user'
+_FLD_TOKEN = 'hshtkn'
 
 
 class IDMappingMongoStorage(_IDMappingStorage):
     """
-    classdocs
+    A MongoDB based implementation of
+    :class:`jgikbase.idmapping.storage.id_mapping_storage.IDMappingStorage`.
+    See that class for method documentation.
     """
 
-    def __init__(self):
+    def __init__(self, db: Database) -> None:
         """
-        Constructor
+        Create a ID mapping storage system.
+
+        :param db: the MongoDB database in which to store the mappings and other data.
         """
+        not_none(db, 'db')
+        self.db = db
+        self._ensure_indexes()
+
+    def _ensure_indexes(self):
+        self.db[_COL_USERS].create_index([(_FLD_AUTHSOURCE, 1), (_FLD_USER, 1)], unique=True)
+        self.db[_COL_USERS].create_index(_FLD_TOKEN, unique=True)
+
+    def create_or_update_local_user(self, user: User, token: HashedToken) -> None:
+        not_none(user, 'user')
+        not_none(token, 'token')
+        try:
+            self.db[_COL_USERS].insert_one({_FLD_AUTHSOURCE: user.authsource.authsource,
```

Comment:
Remember this is an authsource ID, not an authsource. I'll change authsource to authsource_id in a future PR. The class is now AuthsourceID but I forgot to change the fields.

The way this will work is that in the core logic, there will be a way to register an authsource handler. So a request will come in with an authsource id and a token, the core logic will look up a handler with the authsource id and delegate to that handler.